geonode/geoserver/geofence.py

```diff
@@ -0,0 +1,254 @@
+
+        limits = {}
+        for field, value in (
+            ('allowedArea', geo_limit),
+            ('catalogMode', catalog_mode),
+        ):
+            if value is not None:
+                limits[field] = value
+
+        if limits:
+            self.fields['limits'] = limits
+
+    def get_object(self):
+        logger.debug(f"Creating Rule object: {self.fields}")
+        return {'Rule': self.fields}
+
+
+class Batch:
+    def __init__(self, log_name=None) -> None:
+        self.operations = []
+        self.log_name = f'"{log_name}"' if log_name else ''
+
+    def __str__(self) -> str:
+        return super().__str__()
+
+    def add_delete_rule(self, rule_id: int):
+        self.operations.append({
+            '@service': 'rules',
+            '@type': 'delete',
+            '@id': rule_id
+        })
+
+    def add_insert_rule(self, rule: Rule):
+        operation = {
+            '@service': 'rules',
+            '@type': 'insert',
+        }
+        operation.update(rule.get_object())
+        self.operations.append(operation)
+
+    def get_batch_length(self):
+        return len(self.operations)
+
+    def get_object(self):
+        logger.debug(f"Creating Batch object {self.log_name} with {len(self.operations)} operations")
+        return {
+            'Batch': {
+                'operations': self.operations
+            }
+        }
+
+
+class GeofenceClient:
+
+    def __init__(self, baseurl: str, username: str, pw: str) -> None:
+        self.baseurl = baseurl
+        self.username = username
+        self.pw = pw
+
+    def invalidate_cache(self):
+        r = requests.put(
+            f'{self.baseurl}rest/geofence/ruleCache/invalidate',
+            auth=HTTPBasicAuth(self.username, self.pw))
+
+        if r.status_code != 200:
+            logger.warning("Could not invalidate cache")
+            raise GeofenceException("Could not invalidate cache")
+
+    def get_rules(self, page=None, entries=None,
+                  workspace=None, workspace_any=None,
+                  layer=None, layer_any=None):
+        if (page is None and entries is not None) or (page is not None and entries is None):
+            raise GeofenceException(f"Bad page/entries combination {page}/{entries}")
+
+        try:
+            """
+            curl -X GET -u admin:geoserver \
+                http://<host>:<port>/geoserver/rest/geofence/rules.json?page={page}&entries={entries}
+            """
+            params = {}
+
+            if entries:
+                params.update({'page': page, 'entries': entries})
+
+            for param, value in (
+                ('workspace', workspace),
+                ('workspaceAny', workspace_any),
+                ('layer', layer),
+                ('layerAny', layer_any),
+            ):
+                if value is not None:
+                    params[param] = value
+
+            url = f'{self.baseurl}rest/geofence/rules.json?{urllib.parse.urlencode(params)}'
```

Comment:
Can it make sense to handle the baseurl that doesn't end with a `/`? I mean to avoid situations like this:
```
baseurl = "http://localhost/geoserver
.....
URL = "http://localhost/geoserverrest......"
```

Reply:
just used the previous code here